### ISSUE TITLE
A less concise schema file

### ISSUE THREAD

**Issue by matthewmueller** [prisma/specs#339]
In the past we aimed to make the schema easy for developers to understand and edit. Optimizing for reading and writing has led us to support features like implicit many-to-many relations and optional back-relations.

In recent efforts to spruce up our introspection workflow, we've discovered a handful of issues that have led us to step back and reconsider our designs from a more holistic view. 

This issue list the issues we've discovered, explains the shift in perspective, then presents our proposal.

### Problem 1: Implicit Many-to-Many's

Given the following schema,

```groovy
model Student {
  id Int @id
  classes Class[]
}

model Class {
  id Int @id
  students Student[]
}
```

If I migrate, then re-introspect, I end up with a different schema

```groovy
model Student {
  id Int @id
  classes Class[]
}

model _ClassStudent {
  class Class
  student Student
}

model Class {
  id Int @id
  students Student[]
}
```

We can certainly work around this in introspection, and we'll actually have to for our remote migrations table, so this is not a deal breaker.

Where this issue becomes more interesting is when we need to extend our many-to-many with an attribute. For example,

```diff
model Student {
  id Int @id
  classes Class[]
}

model StudentClass {
  class Class
  student Student
+ semester Int
}

model Class {
  id Int @id
  students Student[]
}
```

How does lift handle this transition in a non-destructive way? Are we smart enough to rename the old table to the new table? What happens if the tables are misaligned or one of the relations is forgotten?

We're adding lots of questions (and code) to our plate by offering implicit many-to-manys. This is not to say we can't solve these issues, but at what cost?

### Problem 2: One-to-Ones become One-to-Manys

Given the following schema,

```groovy
model User {
  id Int @id
  post Post
}

model Post {
  user User
}
```

If we migrate with lift, we get,

```sql
create table user (
  id int
)

create table post (
  id int,
  user int not null references user(id)
)
```

Introspecting again results in

```groovy
model User {
  id Int @id
  post Post[]
}

model Post {
  user User
}
```

We can fix this by providing an implicit (or explicit)`@unique` to one side of the relationship. If we go with the **implicit** form, then when we re-introspect, we get:

```groovy
model User {
  id Int @id
  post Post
}

model Post {
  user User @unique
}
```

This is also different than above.

### Problem 3: Required One-to-Ones become Optional on one-side

Starting with our previous example,

```groovy
model User {
  id Int @id
  post Post
}

model Post {
  user User @unique
}
```

If I were to migrate with lift,

```groovy
create table user (
  id int
)

create table post (
  id int,
  user int not null references user(id) unique
)
```

Then re-introspect, my schema will look like this:

```diff
model User {
  id Int @id
- post Post
+ post Post?
}

model Post {
  user User @unique
}
```

This is because two-sided requires are a Prisma-level concept that is not present in any SQL database. There are probably neat tricks we can do in introspection to detect this case, but again at what cost? How far are we willing to go to cover up our warts?

## A Shift in Perspective

There's no great solution here, only tradeoffs. So far we've been steadfast in our desire to make the schema file concise and easy for humans to understand and modify.

What if we loosened this restriction? What if we acknowledged that our domain is varied and complex and that finding a non-leaky, concise syntactical abstraction is too costly for us?

Idan, Johannes and I have been thinking through how we can fix the issues presented above and make schema decisions faster in the future.

We're currently leaning towards treating the schema more like a state file than a schema file. You can think of this similar to the `.git` directory where you rarely edit it directly, but you have tools like the `git` CLI or the Tower UI to interact with it. We'll use the CLI and Studio to primarily interact with our schema file.

On the spectrum of `schema.prisma` as a beautiful centerpiece to `.git` as a hidden directory you never touch, we don't want to go all the way to never-touch-land.

We'd like our schema file to still be understandable and editable, but less concise and beautiful than it currently is. This is to solve problems we have with introspection right now and to unblock future features we want to have like Views. Views may contain large SQL statements that make the `schema.prisma` ugly.

Our shift in perspective is that we'd now be okay with this ugliness and use tooling to make interacting with this file easier.

## Proposal

Hopefully I've explained our shift in perspective. Let's hop off the product clouds and bring this back to earth. What does this mean for the `schema.prisma` today?

### Remove implicit many-to-many relations

Today we support the implicit many-to-many's below,

```groovy
model Student {
  id Int @id
  classes Class[]
}

model Class {
  id Int @id
  students Student[]
}
```

Instead of supporting this, I propose we require the more explicit form

```groovy
model Student {
  id Int @id
}

model ClassStudent {
  student Student
  class Class
}

model Class {
  id Int @id
}
```

This closely models SQL, so lift, then introspection will produce the same schema. You'll notice I also removed the back-relations, I'll explain this next.

### Remove back-relations. Configure the back-relation on the forward-relation

To remove some of the ambiguities presented above, we're thinking of removing the back-relations and letting you configure the back-relation on the forward-relation. Let's see how this looks:

**One-to-One**

```groovy
// before
model User {
  id Int @id
  customer Customer
}
model Customer {
  user User
}

// after
model User {
  id Int @id
}
model Customer {
  user User @unique
}
```

**One-to-Many**

```groovy
// before
model User {
  id Int @id
  posts Post[]
}
model Post {
  user User
}

// after
model User {
  id Int @id
}
model Post {
  user User
}
```

**Many-to-Many (you saw this one above)**

```groovy
// before
model Student {
  id Int @id
  classes Class[]
}
model Class {
  id Int @id
  students Student[]
}

// after
model Student {
  id Int @id
}
model ClassStudent {
  student Student
  class Class
}
model Class {
  id Int @id
}
```

By removing the back-relation we've lost the ability to customize our backrelation field name for Photon. We can add that back by adding an attribute to the forward relation. It's not pretty, but it'll get the job done:

```groovy
model User {
  id Int @id
}
model Post {
  user User @backrelation("posts")
}
```

I've named this attribute `@backrelation`, but there's probably a better name.

### Closing Thoughts

Undoubtably, we've made our schema file worse. Alice would not be happy. However, the result of this change is a more explicit and sound structure. We're following the well-trodden footsteps of data modeling and normalization.

To satisfy Alice's needs, we'll build tooling (e.g. Prisma CLI) and UI (e.g. Prisma Studio) to interact with our schema file. Alice should never need to edit this file if she doesn't want to.

**Comment by sorenbs** [comment]
I really enjoyed reading this well-articulated issue. The fact that you laid out motivating problems first enable me to comment on the problems and proposed solution separately, hopefully enabling a more nuanced conversation.

Following are my comments to each section in this issue. They are best read by first reading each individual section in the issue.

### Problems

- **Problem 1: Implicit Many-to-Many's**
    - This is a solved problem:
    - Introspection should follow our conventions when printing schema
    - VS Code plugin should support converting from implicit to explicit, making it easy to add a field to the relation table
- **Problem 2: One-to-Ones become One-to-Manys**
    - Introspection should take presence of a unique constraint into account when deciding between 1:1 and 1:m. The unique constraint should not be written in the schema for 1:1 relations because it is implied (otherwise it would be a 1:m relation).
- **Problem 3: Required One-to-Ones become Optional on one-side**
    - I am in favour of not supporting double-required one-one for this reason. We should disallow this at any time we are parsing the schema: engine startup, in VS Code extension, Studio, Lift up.
    - In the future, we could decide to implement this in the database using `DEFERRABLE` in pg, SQLite and Oracle. MySQL and MsSQL does not support it.
        - https://www.cybertec-postgresql.com/en/1-to-1-relationship-in-postgresql-for-real/
        - https://sequelize.readthedocs.io/en/v3/api/deferrable/
    - In the future we could decide to implement this on the Query Engine level in order to support it for MySQL, MsSQL and Mongo
- **Views may contain large SQL statements that make the schema.prisma ugly**
    - This is not a problem. We have discussed ways to split the schema file into multiple files, which neatly solves this issue

### Proposal

- **Remove implicit many-to-many relations**
    - When just viewed from the SQL perspective, this makes sense. If we go this route, we should ensure that whenever we parse the schema, we should detect the implicit case and give the user the correct representation.
    - The problem is that this does not fit the Mongo model. In Mongo we model m-n relations as two embedded arrays of IDs, which performs much better than a relation collection, as Mongo does not support joins.
- **Remove back-relations. Configure the back-relation on the forward-relation**
    - I don't think this change is required in order to address the motivating issues. From an implementation perspective we don't gain anything from going this route.
    - We have discussed this at length in the past. We landed on the current approach after having had both implemented in code that shipped to users. I don't see any new data that justifies this change.
    - Tangentially related, I would be in favour of finding a way to avoid the relation name. In order to solve the ambiguous relations case, @do4gr suggested we could name the opposite field instead of introducing relation names.

### Closing Thoughts

- If the user is to primarily interact with the schema through tooling, then why do we care about these changes?

**Comment by matthewmueller** [comment]
Answers inline!

**Introspection should follow our conventions when printing schema**

Same as next answer

**Introspection should take presence of a unique constraint into account when deciding between 1:1 and 1:m. The unique constraint should not be written in the schema for 1:1 relations because it is implied (otherwise it would be a 1:m relation).**

I'm under the impression that this requires more code & testing to do this. If I'm wrong about this, then we're in agreement.

Taking the side "If the user is to primarily interact with the schema through tooling, then why do we care about these changes?" 

Then we should opt for less internal code because we no longer care about making the schema pretty.

**I am in favour of not supporting double-required one-one for this reason. We should disallow this at any time we are parsing the schema: engine startup, in VS Code extension, Studio, Lift up.**

I'm also in favor of this, at least initially. This is a Prisma-level constraint that we can add later in a non-breaking way (if it's a compiler error initially). @schickling is championing this concept.

**This [large SQL statements in views] is not a problem. We have discussed ways to split the schema file into multiple files, which neatly solves this issue**

**The problem is that this does not fit the Mongo model. In Mongo we model m-n relations as an embedded array of IDs, which performs much better than a relation collection, as Mongo does not support joins.**

**Comment by sorenbs** [comment]
The changes discussed here are superseded by the new proposal in https://github.com/prisma/specs/issues/348
